### PR TITLE
Add constraint for tags

### PR DESCRIPTION
**Pull request by drvinceknight**
Also:

- Added sessions in the test fixtures.
- Added some specific numpy tests that check that the constraints work as expected.
- Modified `parameters.constraints` to deal with constraints that take different arguments.

### CODE REVIEW

File: conference_scheduler/tests/test_parameters.py [hunk 1]
```diff
@@ -10,12 +21,86 @@ def test_schedule_all_events(shape, X):
     constraints = [c for c in parameters._schedule_all_events(shape, X)]
     assert len(constraints) == 3
 
+def test_schedule_all_events_fails_np(shape):
+    # Third talk is not scheduled
+    X = np.array([[1, 0, 0, 0, 0, 0, 0],
+                  [0, 0, 0, 0, 1, 0, 0],
+                  [0, 0, 0, 0, 0, 0, 0]])
+    constraint = all(parameters._schedule_all_events(shape, X))
```
Comment (meatballs) [inline conference_scheduler/tests/test_parameters.py]: nice! Much better

File: conference_scheduler/parameters.py [hunk 2]
```diff
@@ -21,11 +50,45 @@ def _max_one_event_per_slot(shape, X):
...
+        for slot, event in itertools.product(slots, event_indices):
+            other_events = _events_with_diff_tag(event, tag_array)
+            for other_slot, other_event in itertools.product(slots, other_events):
+                if other_slot != slot and other_event != event:
+                    # If they have different tags they cannot be scheduled
+                    # together
+                    yield X[(event, slot)] + X[(other_event, other_slot)] <= 1
+
+
+def constraints(shape, session_array, tag_array, X):
```
Comment (meatballs) [inline conference_scheduler/parameters.py]: Nitpicking: Given that the constraint functions all have `shape, X` within their signature, could we keep the order of the parameters to always start with those two?

Comment (drvinceknight) [inline conference_scheduler/parameters.py]: Sure happy to reorder although `_events_in_sessions_share_a_tag` doesn't take `shape` as an argument (just `session_array` and `tag_array`): would you still like me to tweak it?

Here's what the `arg` iterates through:

```python
 90     args = ((shape, ), (shape, ), (session_array, tag_array)) 
```

So all constraints will take one of those and `X`.

Comment (drvinceknight) [inline conference_scheduler/parameters.py]: Not entirely unrelated:

> Here's what the arg iterates through:

I've just pushed e63d831 which changes that to be kwargs:

```python
90     generator_kwargs = ({"shape":shape}, {"shape":shape},                                                                                          91                         {"session_array": session_array, "tag_array": tag_array}) 
```

Happy to revert.

**Comment by drvinceknight** [comment]
13a477d tweaks the travis build so that it automatically installs the necessary library. (Previous one failed because of no `requirements.txt`.)

File: conference_scheduler/parameters.py (hunk 2 above)
Comment (meatballs) [inline conference_scheduler/parameters.py]: I reckon using kwargs is a much better solution.

File: conference_scheduler/parameters.py [hunk 3]
```diff
@@ -21,11 +50,46 @@ def _max_one_event_per_slot(shape, X):
...
     generators = (
         _schedule_all_events,
         _max_one_event_per_slot,
+        _events_in_session_share_a_tag,
     )
-    for generator in generators:
-        for constraint in generator(shape, X):
+    generator_kwargs = ({"shape":shape}, {"shape":shape},
```
Comment (meatballs) [inline conference_scheduler/parameters.py]: shouldn't the second one here be X rather than shape?

Comment (meatballs) [inline conference_scheduler/parameters.py]: forget that. just read further down the code!

Comment (meatballs) [inline conference_scheduler/parameters.py]: but why is the second `shape` their at all?

Comment (drvinceknight) [inline conference_scheduler/parameters.py]: We have 3 constraint generators:

```python
 85     generators = (                                                                                                                                
 86         _schedule_all_events,                                                                                                                     
 87         _max_one_event_per_slot,                                                                                                                  
 88         _events_in_session_share_a_tag,                                                                                                           
 89     )    
```

Each of those is going to be called with (potentially different) kwargs:

```
 90     generator_kwargs = ({"shape":shape}, {"shape":shape},                                                                                         
 91                         {"session_array": session_array, "tag_array": tag_array}) 
```

That is done by zipping the two tuples together:

```python
93     for generator, kwargs in zip(generators, generator_kwargs):
```

So we need `shape` twice: for the first two generators.

Comment (meatballs) [inline conference_scheduler/parameters.py]: Duh!

Sorry, not had coffee yet this morning!

**Pull request merged**